sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java

```diff
@@ -217,26 +218,39 @@ protected AvroCoder(Class<T> type, Schema schema) {
     this.decoder = new EmptyOnDeserializationThreadLocal<>();
     this.encoder = new EmptyOnDeserializationThreadLocal<>();
 
-    // Reader and writer are allocated once per thread and are "final" for thread-local Coder
-    // instance.
-    this.reader = new EmptyOnDeserializationThreadLocal<DatumReader<T>>() {
+    // ReflectData, Reader and writer are allocated once per thread per Coder
+    this.reflectData = new EmptyOnDeserializationThreadLocal<ReflectData>() {
```

Comment:
No point of having a separate thread local for the ReflectData. We should attempt to save an instance per avro coder if we believe that this will clean up the class leak in Avro. Use Guavas memoizing supplier with a function that returns a new instance of ReflectData. Guavas memoizing supplier cache is transient.

The decoder/encoder/writer/reader aren't threadsafe which is why we have thread locals for them.